llama-index-integrations/llms/llama-index-llms-lmstudio/llama_index/llms/lmstudio/base.py

```diff
@@ -0,0 +1,245 @@
+from llama_index.core.llms.custom import CustomLLM
+
+from llama_index.core.base.llms.generic_utils import (
+    stream_chat_to_completion_decorator,
+    chat_to_completion_decorator,
+    achat_to_completion_decorator,
+    astream_chat_to_completion_decorator,
+)
+from llama_index.core.constants import (
+    DEFAULT_TEMPERATURE,
+    DEFAULT_CONTEXT_WINDOW,
+    DEFAULT_NUM_OUTPUTS,
+)
+from llama_index.core.base.llms.types import (
+    ChatMessage,
+    ChatResponse,
+    ChatResponseGen,
+    CompletionResponse,
+    CompletionResponseAsyncGen,
+    CompletionResponseGen,
+    LLMMetadata,
+    MessageRole,
+)
+
+
+def get_additional_kwargs(
+    response: Dict[str, Any], exclude: Tuple[str, ...]
+) -> Dict[str, Any]:
+    return {k: v for k, v in response.items() if k not in exclude}
+
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_REQUEST_TIMEOUT = 30.0
+
+
+class LMStudio(CustomLLM):
+    base_url: str = Field(
+        default="http://localhost:1234/v1",
+        description="Base url the model is hosted under.",
+    )
+
+    context_window: int = Field(
+        default=DEFAULT_CONTEXT_WINDOW,
+        description="The maximum number of context tokens for the model.",
+        gt=0,
+    )
+
+    model_name: str = Field(description="The model to use.")
+
+    request_timeout: float = Field(
+        default=DEFAULT_REQUEST_TIMEOUT,
+        description="The timeout for making http request in seconds to LM Studio API server.",
+    )
+    num_output: int = Field(
+        default=DEFAULT_NUM_OUTPUTS,
+        description=LLMMetadata.__fields__["num_output"].field_info.description,
+    )
+
+    is_chat_model: bool = Field(
+        default=True,
+        description=(
+            "LM Studio API supports chat."
+            + LLMMetadata.__fields__["is_chat_model"].field_info.description
+        ),
+    )
+
+    temperature: float = Field(
+        default=DEFAULT_TEMPERATURE,
+        description=("The temperature to use for sampling."),
+        gte=0.0,
+        lte=1.0,
+    )
+
+    timeout: float = Field(
+        default=120, description=("The timeout to use in seconds."), gte=0
+    )
+
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict, description=("Additional kwargs to pass to the model.")
+    )
+
+    def __create_payload_from_messages(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> Dict[str, Any]:
+        return {
+            "model": self.model_name,
+            "messages": [
+                {
+                    "role": message.role.value,
+                    "content": message.content,
+                    **(
+                        message.additional_kwargs
+                        if message.additional_kwargs is not None
+                        else {}
+                    ),
+                }
+                for message in messages
+            ],
+            "options": self._model_kwargs,
+            "stream": False,
```

Comment:
stream is hardcoded to false here (I think we want to pass this in?)

Reply:
It is not actually hardcoded. We can override the value by passing an argument in the function. In the returned dictionary, **kwargs will include all keyword arguments that are not explicitly listed in the function signature. If any of these keyword arguments match the keys in the dictionary, they will override the hardcoded values.